packages/react-positioning/src/usePopper.ts

```diff
@@ -298,86 +110,103 @@ export function usePopper(
-  //
-  // The first scenario are updates for a targetRef, we can handle it properly only via callback refs, but it causes
-  // re-renders (we would like to avoid them).
-  //
-  // The second problem is related to refs resolution on React's layer: refs are resolved in the same order as effects
-  // that causes an issue when you have a container inside a target, for example: a menu in ChatMessage.
-  //
-  // function ChatMessage(props) {
-  //   <div className="message" ref={targetRef}> // 3) ref will be resolved only now, but it's too late already
-  //     <Popper target={targetRef}> // 2) create a popper instance
-  //       <div className="menu" /> // 1) capture ref from this element
-  //     </Popper>
-  //   </div>
-  // }
-  //
-  // Check a demo on CodeSandbox: https://codesandbox.io/s/popper-refs-emy60.
-  //
-  // This again can be solved with callback refs. It's not a huge issue as with hooks we are moving popper's creation
-  // to ChatMessage itself, however, without `useCallback()` refs it's still a Pandora box.
-  const targetRef = useCallbackRef<HTMLElement | PopperJs.VirtualElement | null>(null, handlePopperUpdate, true);
-  const containerRef = useCallbackRef<HTMLElement | null>(null, handlePopperUpdate, true);
-  const arrowRef = useCallbackRef<HTMLElement | null>(null, handlePopperUpdate, true);
-
-  // Stores external target from options.target or setTarget
-  const overrideTargetRef = useCallbackRef<HTMLElement | PopperVirtualElement | null>(null, handlePopperUpdate, true);
+  const updatePosition = React.useState(() => debounce(forceUpdate))[0];
+
+  const targetRef = useCallbackRef<HTMLElement | PopperVirtualElement | null>(null, (target, prevTarget, isFirst) => {
+    addEventListenerOnScrollParent(target, prevTarget, updatePosition);
+
+    if (!isFirst) {
+      updatePosition();
+    }
+  });
+  const containerRef = useCallbackRef<HTMLElement | null>(null, (container, prevContainer, isFirst) => {
+    if (container && enabled) {
+      // When the container is first resolved, set position `fixed` to avoid scroll jumps.
+      // Without this scroll jumps can occur when the element is rendered initially and receives focus
+      Object.assign(container.style, { position: 'fixed', top: 0, left: 0 });
+    }
+
+    addEventListenerOnScrollParent(container, prevContainer, updatePosition);
+
+    if (!isFirst) {
+      updatePosition();
+    }
+  });
+  const arrowRef = useCallbackRef<HTMLElement | null>(null, updatePosition, true);
+  const overrideTargetRef = useCallbackRef<HTMLElement | PopperVirtualElement | null>(
+    null,
+    (target, prevTarget, isFirst) => {
+      addEventListenerOnScrollParent(target, prevTarget, updatePosition);
+
+      if (!isFirst) {
+        updatePosition();
+      }
+    },
+  );
 
   React.useImperativeHandle(
     options.popperRef,
     () => ({
-      updatePosition: () => {
-        popperInstanceRef.current?.update();
-      },
+      updatePosition,
```

Comment:
Can we avoid leaking `Promise` into public API? As it might be confusing for customers:

![image](https://user-images.githubusercontent.com/14183168/156767380-5f2eeb81-da88-4f9a-837f-0a0868c8a056.png)

Reply:
The types don't leak to the public API because the `PopperRefHandle` is typed correctly